Re-enable Snapshots tab

glekner opened the pull request.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-details-page.tsx:
```diff
@@ -41,6 +49,8 @@ export const breadcrumbsForVMPage = (match: any) => () => [
 export const VirtualMachinesDetailsPage: React.FC<VirtualMachinesDetailsPageProps> = (props) => {
   const { name, ns: namespace } = props.match.params;
+  const [snapData, snapLoaded, snapErr] = useK8sGet(VirtualMachineSnapshotModel);
+  const isSnapshotSupported = snapData && snapLoaded && !snapErr;
```
can you please look at the original description of this bug https://bugzilla.redhat.com/show_bug.cgi?id=1881125 ? And fix the loading when reintroducing the snapshots again?

glekner commented:
@suomiy we know look for snapshots only in the relevant namespace. Is this solution sufficient?

suomiy commented:
@glekner we should check if the CRD is present

pcbailey commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-details-page.tsx:
```diff
@@ -22,10 +23,21 @@ import { VMConsoleFirehose } from './vm-console';
 import { VMNics } from '../vm-nics';
 import { PendingChangesWarningFirehose } from './pending-changes-warning';
+import { VMSnapshotsPage } from '../vm-snapshots/vm-snapshots';
+import {
```
nitpick: This should be above the kubevirt package imports. Of course, we can catch it in a follow-up if you don't have any other changes to make.

glekner commented:
added `useModelLoaded` https://github.com/openshift/console/pull/6807/files#diff-1c59bec23f195454582776c62b6085baR64
@suomiy @rawagner

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-details-page.tsx:
```diff
@@ -41,6 +49,7 @@ export const breadcrumbsForVMPage = (match: any) => () => [
 
 export const VirtualMachinesDetailsPage: React.FC<VirtualMachinesDetailsPageProps> = (props) => {
   const { name, ns: namespace } = props.match.params;
+  const isSnapshotsAvailable = useModelLoaded(VirtualMachineSnapshotModel.kind);
```
shouldn't `referenceForModel` be used here?

glekner replied:
depends on the use case, some would use `referenceForModel(model)` but in our case the resource in store is called `VMSnapshotModel` so we need `VMSnapshotModel.kind`

suomiy replied:
it might get loaded only with the group+version to the store so we should use `referenceForModel` to be safe

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-details-page.tsx:
```diff
@@ -95,6 +102,7 @@ export const VirtualMachinesDetailsPage: React.FC<VirtualMachinesDetailsPageProp
     consolePage,
     nicsPage,
     disksPage,
+    ...(isSnapshotsAvailable ? [snapshotsPage] : []),
```
are we expecting more pages? IMO simple if would be more readable.

glekner replied:
this is a one line and it is a convention.
check https://github.com/openshift/console/blob/master/frontend/webpack.config.ts#L192 for example

suomiy replied:
I wouldn't cite webpack.config.ts as a convention. I know there are both use cases in the code base.

This one is less elegant. 
Do as you want - I can live with both.

suomiy commented on frontend/public/components/utils/k8s-watch-hook.ts:
```diff
@@ -61,6 +61,9 @@ const useModelsLoaded = (): boolean => {
 };
 
+export const useModelLoaded = (model: string): boolean =>
+  !!useSelector<RootState, K8sKind>(({ k8s }) => k8s.getIn(['RESOURCES', 'models', model]));
```
we could check for empty `model` first

glekner replied:
react hooks shouldn't be called conditionally. eslint will shout at you

suomiy replied:
yes, unless the condition is inside

suomiy commented on frontend/public/components/utils/k8s-watch-hook.ts:
```diff
@@ -61,6 +61,9 @@ const useModelsLoaded = (): boolean => {
   return ref.current;
 };
 
+export const useModelLoaded = (model: string): boolean =>
```
+1 hook might be useful

rawagner replied:
You can use https://github.com/openshift/console/blob/master/frontend/packages/console-shared/src/hooks/useK8sModel.ts instead

suomiy replied:
yeah that sound much better

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshots.tsx:
```diff
@@ -91,7 +98,22 @@ export const VMSnapshotsPage: React.FC<VMLikeEntityTabProps> = ({ obj: vmLikeEnt
   );
 
-  const [snapshots, snapshotsLoaded, snapshotsError] = useK8sWatchResource<VMSnapshot[]>(resource);
+  const restoreResource: WatchK8sResource = React.useMemo(
```
useK8sWatchResource doesnt require memoized resource anymore :)

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/snapshot-row.tsx:
```diff
@@ -66,12 +56,14 @@ const getActions = (
   }
 
-  const actions = [menuActionRestore, menuActionDelete];
+  const actions = [menuActionDelete];
```
what about the snapshot detail later?

glekner replied:
snapshot detail?
the restore is now a dedicated button for each row, not a kebab option

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/utils.ts:
```diff
@@ -4,6 +4,8 @@ import { Kebab } from '@console/internal/components/utils';
 export const snapshotsTableColumnClasses = [
```
the total is 12

glekner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshots.tsx:
```diff
@@ -64,6 +64,13 @@ export const VMSnapshotsTable: React.FC<VMSnapshotsTableProps> = ({
             transforms: [sortable],
           },
+          {
+            title: 'Last restored',
```
@suomiy 
since the latest restore date is computed inside the row component (because one snapshot can have multiple restore objects) we can't sort the table by it. do you have any suggestions?

https://github.com/openshift/console/pull/6807/files#diff-d4bb250ed498bbf43eeee21385193b71R146

rawagner replied:
@glekner you can use sortFunc prop

glekner replied:
done, hope it won't cause performance issues
https://github.com/openshift/console/pull/6807/files#diff-b2d4cf524cc534d555f71cccff72c6ecR172

suomiy replied:
I am not a big fan of expecting that the core code should handle our CRDs. Wouldn't it be more straightforward to just resolve the time here and pass it as a data like we do in our other tables?

glekner commented:
Updated to match latest design

Restore failed popover:
<img width="812" alt="Screen Shot 2020-10-12 at 13 03 24" src="https://user-images.githubusercontent.com/24938324/95746073-def69480-0c9e-11eb-9b66-08c3cc75cbcc.png">

Restore in progress popover:
<img width="774" alt="Screen Shot 2020-10-12 at 15 18 12" src="https://user-images.githubusercontent.com/24938324/95746100-e74ecf80-0c9e-11eb-835b-056d41e4900d.png">

can you review @suomiy @matthewcarleton @yfrimanm

yfrimanm commented:
Thanks @glekner 
Nit:  For the restoring popover, maybe we can make it clearer with: ''Virtual-machine-name 2019-8-28 is being restored from this snapshot''.

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/types.ts:
```diff
@@ -6,4 +7,6 @@ export type VMSnapshotRowCustomData = {
   columnClasses: string[];
   isDisabled: boolean;
+  restores: VMRestore[];
+  withProgress: (promise: Promise<any>) => void;
```
why do we need withProgress in snapshots?

glekner replied:
why not? its a resource creation like any other modal?

suomiy replied:
because this feature is used only by the wizard

suomiy replied:
ah you are right, It is used by the detail pages as well

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshots.tsx:
```diff
@@ -81,17 +89,33 @@ export const VMSnapshotsPage: React.FC<VMLikeEntityTabProps> = ({ obj: vmLikeEnt
+  const [restores, restoresLoaded, restoresError] = useK8sWatchResource<VMRestore[]>(
+    restoreResource,
+  );
+  const sortedRestores = React.useMemo(
```
it might be easier to create a map with `getVmRestoreSnapshotName` key here for better consumption. 
Then it wouldn't need to be even sorted as you could just save the relevant restore.

glekner replied:
which snapshotName? one vm can have multiple snapshots with different names? a row for each.

suomiy replied:
yes, but relevant restore shoul be only one , right?

glekner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/snapshot-row.tsx:
```diff
@@ -66,32 +62,103 @@ const getActions = (
+  const snapshotReady = isVMSnapshotReady(snapshot);
+
+  if (snapshotError) {
+    return <ErrorStatus title={'Create Failed'}>{snapshotError?.message}</ErrorStatus>;
```
user can access the snapshot details by clicking on its name in the table

glekner commented:
- mapped restores by key
- changed kebab options to: 
1. Last Restore details (if available)
2. Delete Snapshot
![image](https://user-images.githubusercontent.com/24938324/96141056-4ca5f900-0f09-11eb-9789-c2520e8c34f3.png)

@suomiy can you review?
the common kebab options can be confusing here because of the "Dual" resource page of snapshots/kebab. if a user wants to play with them, they should visit a specific resource page

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshots.tsx:
```diff
@@ -81,17 +93,46 @@ export const VMSnapshotsPage: React.FC<VMLikeEntityTabProps> = ({ obj: vmLikeEnt
+
+  const mappedRelevantRestores = React.useMemo(
+    () =>
+      restores.reduce((restoreMap, currentRestore) => {
```
this would be simpler

```tsx
      restores.reduce((restoreMap, currentRestore) => {
        const relevantRestore = restoreMap[getVmRestoreSnapshotName(currentRestore)];
        if (!relevantRestore || new Date(getVmRestoreTime(relevantRestore)).getTime() < new Date(getVmRestoreTime(currentRestore)).getTime()) {
            restoreMap[getVmRestoreSnapshotName(currentRestore)] = currentRestore
        }
        return restoreMap
      }, {}),
```

suomiy commented:
> the common kebab options can be confusing here because of the "Dual" resource page of snapshots/kebab. if a user wants to play with them, they should visit a specific resource page

I was under an impression that it would make more sense to show snapshots as one logical resources (ie as an aggregation of resources like vm/vmi does). 
Do you think it makes sense to introduce VMRestore resource to the users? And what advantages does it have?

glekner commented:
* Created custom Details Page for VM Snapshot with Description if available
* Added Restore action, Status and Last Restore(same new component like the table)
* Removed Last Restore Details kebab option in a Snapshot row 
* Added breadcrumbs for VM context

<img width="832" alt="Screen Shot 2020-10-20 at 21 03 00" src="https://user-images.githubusercontent.com/24938324/96626550-50c28400-1318-11eb-9d4c-8b0258792f6e.png">

@suomiy @matthewcarleton @yfrimanm can you take a look

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/snapshot-details.tsx:
```diff
@@ -0,0 +1,171 @@
+
+  const [restores, restoresLoaded] = useK8sWatchResource<VMRestore[]>(restoreResource);
+
+  const mappedRelevantRestores = React.useMemo(
```
could we make a hook for this, since we are using ti twice?

rawagner replied:
you could include `useK8sWatchResource` into the new hook and it would return `[mappedRestores, loaded, error]`

suomiy reviewed (changes requested):
@glekner we probably should ask for delete confirmation of snapshots when we delete the VM. The same we do with disks. To prevent stale snapshots being stored in the system. 
Thoughts?

Also not sure if we want an owner reference on them.

glekner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/snapshot-row.tsx:
```diff
@@ -15,38 +18,30 @@ import {
-    ),
-});
 
 const menuActionDelete = (
```
We don't need any of it actually, the only kebab option as of now is Delete.so now just using 
`const { Delete } = Kebab.factory`

glekner commented:
@suomiy @rawagner 
* Implemented suggestions
* Created `useMappedVMRestores` hook
* Made `DescriptionModal` generic to reuse in the snapshot details.

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/use-mapped-vm-restores.ts:
```diff
@@ -0,0 +1,41 @@
+    [restores],
+  );
+
+  return [mappedRelevantRestores, restoresLoaded, restoresError];
```
you can return directly from React.useMemo so you wont create a new array every time

glekner replied:
fixed

suomiy commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshot-details.tsx:
```diff
@@ -0,0 +1,157 @@
+      },
+      {
+        name: `${vmName} Snapshots`,
+        path: `/k8s/ns/${match.params.ns ||
```
can use the variable here as well

glekner replied:
fixed

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/description-modal/description-modal.tsx:
```diff
@@ -8,28 +8,26 @@ import {
 
-export const VMDescriptionModal = withHandlePromise((props: VMDescriptionModalProps) => {
-  const { vmLikeEntity, inProgress, errorMessage, handlePromise, close, cancel } = props;
+const DescriptionModal = withHandlePromise((props: DescriptionModalProps) => {
```
```suggestion
const DescriptionModal = withHandlePromise<DescriptionModalProps>((props) => {
```

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/description-modal/index.ts:
```diff
@@ -0,0 +1 @@
+export * from './description-modal';
```
lets avoid creating new index files, they cause cyclic deps and we will be removing them in the future

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/snapshot-modal/snapshot-modal.tsx:
```diff
@@ -58,6 +60,13 @@ const SnapshotsModal = withHandlePromise((props: SnapshotsModalProps) => {
+          <FormRow title="Description" fieldId={asId('desc')}>
+            <TextArea
+              value={description}
+              onChange={(d) => setDescription(d)}
```
```suggestion
              onChange={setDescription}
```

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-snapshots/vm-snapshot-details.tsx:
```diff
@@ -0,0 +1,156 @@
+      resources={[resource]}
+      menuActions={menuActions}
+      pages={pages}
+      breadcrumbsFor={(obj) => breadcrumbsForSnapshots(obj)}
```
```suggestion
      breadcrumbsFor={breadcrumbsForSnapshots}
```

rawagner commented on frontend/packages/kubevirt-plugin/src/selectors/selectors.ts:
```diff
@@ -26,7 +25,7 @@ export const getLabelValue = (entity: K8sResourceKind, label: string): string =>
 // Annotations
 export const getAnnotations = (
-  vm: VMGenericLikeEntityKind,
+  vm: K8sResourceKind,
```
```suggestion
  vm: K8sResourceCommon,
```

rawagner commented:
/lgtm
